### PR TITLE
[COT] `query()` method: add support for field queries

### PR DESCRIPTION
**Pull request by jorgeatorres**
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds support for a `field_query` arg to `OrdersTableQuery`, which can be used via `wc_get_orders()` when COT is active.
This came to be as a result of an internal discussion (p1660893943441389-slack-C9QDV4TL4) where we wanted to add a way to easily query orders matching various fields and value combinations without the need to write custom SQL queries. This wasn't really possible before, but thanks to most of our fields being metadata, `WP_Query`'s `meta_query` could be used for that purpose. As such, `field_query` supports most (or all?) of `meta_query`'s (metavalue) functionality but operates on core order fields instead of metadata.

With the new `field_query` you can perform complex queries: matching various fields and with AND/OR relations. For example, to obtain all pending/on-hold orders where the amount > 10.0 and the purchaser's first name or last name contains "michel" you could do this:

```php
$orders = wc_get_orders(
	[
		'field_query' => [
			[
				'field'   => 'total',
				'value'   => '10.0',
				'compare' => '>',
				'type'    => NUMERIC,
			],
			[
				'relation' => 'OR',
				[
					'field'   => 'billing_first_name',
					'value'   => 'michel',
					'compare' => 'LIKE',
				],
				[
					'field'   => 'billing_last_name',
					'value'   => 'michel',
					'compare' => 'LIKE',
				]
			]
		],
		'status' => [ 'wc-pending', 'wc-on-hold' ]
	]
);
```

The syntax for `field_query` clauses is analogous to that of WP's [meta query](https://developer.wordpress.org/reference/classes/wp_meta_query/) but with `field` standing for `key`.

Also, `field_query` can be combined with other types of queries already supported by `OrdersTableQuery` such as direct access to certain fields, `meta_query` and `date_query`.

#### Option 1
1. Check added unit tests.

#### Option 2
1. Activate and make COT authoritative.
2. Populate the db with some orders and data.
3. Run your own queries using `wc_get_orders()`. Join the fun.

### Next Steps
With `field_query` we can rewrite some of the code in `OrdersTableQuery`, dealing with basic columns in the core tables, in a more concise manner. My plan is to do that in a subsequent PR, as well adding an abstract class for this kind of "recursive" query to share code between `OrdersTableMetaQuery` and `OrdersTableFieldQuery`. We've also discussed the possibility of supporting these queries in the CPT implementation to make things simpler for devs.

Once this PR is reviewed/merged I'll submit PRs for the above.

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [X] Have you written new tests for your changes, as applicable?
-   [X] Have you successfully run tests with your changes locally?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+			case 'NOT IN':
+				$where = $wpdb->prepare( '(' . substr( str_repeat( ',%s', count( $clause_value ) ), 1 ) . ')', $clause_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+				break;
+			case 'BETWEEN':
+			case 'NOT BETWEEN':
+				$where = $wpdb->prepare( '%s AND %s', $clause_value[0], $clause_value[1] );
+				break;
+			case 'LIKE':
+			case 'NOT LIKE':
+				$where = $wpdb->prepare( '%s', '%' . $wpdb->esc_like( $clause_value ) . '%' );
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php]: seems like `$clause_compare` is not SQL escaped for these cases, we should escape them.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php [hunk 2]
```diff
@@ -604,6 +619,55 @@ private function build_query(): void {
...
+		}
+
+		$is_address_mapping = in_array( $mapping_id, array( 'billing_address', 'shipping_address' ), true );
+
+		$alias   = $this->get_core_mapping_alias( $mapping_id );
+		$table   = $is_address_mapping ? $this->tables['addresses'] : $this->tables[ $mapping_id ];
+		$join    = '';
+		$join_on = '';
+
+		$join .= "INNER JOIN {$table}" . ( $alias !== $table ? " AS {$alias}" : '' );
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableQuery.php]: table name should be escaped in backticks

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php [hunk 3]
```diff
@@ -995,4 +995,104 @@ public function test_cot_query_search(): void {
 			'Search terms match against address data as well as order item names.'
 		);
 	}
+
+	/**
+	 * Ensure field_query works as expected.
+	 */
+	public function test_cot_query_field_query(): void {
```
Comment (vedanshujain) [inline plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php:1093]: Can we add more tests where
1. we do field along with other queries (meta, ids etc) at the same time
2. query result is empty
3. invalid value types
4. invalid column names

**Review by vedanshujain — COMMENTED** [review]
I'm still reviewing, but pretty neat work, as usual. I wonder, can we add a compatibility layer where we also read meta_queries, and if a meta_key is the same as a column name then we implicitly convert it into a field query? That way, existing meta_queries will work seamlessly on COT as well without needing any updation.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php (hunk 1 above)
Comment (jorgeatorres) [reply plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php]: @vedanshujain: could you let me know which cases need to be SQL-escaped? I might be missing something, but as far as I can see, all are using `%s` placeholders.

File: plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php (hunk 3 above)
Comment (jorgeatorres) [reply plugins/woocommerce/tests/php/src/Internal/DataStores/Orders/OrdersTableDataStoreTests.php:1093]: I've added a few more tests, but please let me know if you want to see something else in particular.

**Comment by jorgeatorres** [comment]
Hey @vedanshujain!

Thanks for the review. I've rebased to fix conflicts, and added some more validation and tests to the code. Please let me know what you think.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php (hunk 1 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php]: You are right, they are indeed prepared. Not sure why I said that, thanks!

**Comment by vedanshujain** [comment]
Looks great, thanks for your amazing work both here and with OrderTableQuery 🙌

**Review by vedanshujain — APPROVED** [review]

**Pull request merged**